File: flutter_client/lib/core/models/corporation_profile.dart
class CorporationProfile {
  final CorporationIdentity identity;
  final CorporationMembershipProfile membership;
  final CorporationCapacityProfile capacity;
  final CorporationCapacityFinance capacityFinance;
  final CorporationAccounts accounts;
  final CorporationFiscalProfile fiscal;
  final CorporationTechnologyProfile technology;
  final CorporationGovernanceProfile governance;
  final CorporationConstitutionProfile constitution;

  const CorporationProfile({
    required this.identity,
    required this.membership,
    required this.capacity,
    required this.capacityFinance,
    required this.accounts,
    required this.fiscal,
    required this.technology,
    required this.governance,
    required this.constitution,
  });

  factory CorporationProfile.fromJson(Map<String, dynamic> json) =>
      CorporationProfile(
        identity: CorporationIdentity.fromJson(_map(json['identity'])),
        membership:
            CorporationMembershipProfile.fromJson(_map(json['membership'])),
        capacity: CorporationCapacityProfile.fromJson(_map(json['capacity'])),
        capacityFinance:
            CorporationCapacityFinance.fromJson(_map(json['capacityFinance'])),
        accounts: CorporationAccounts.fromJson(_map(json['accounts'])),
        fiscal: CorporationFiscalProfile.fromJson(_map(json['fiscal'])),
        technology:
            CorporationTechnologyProfile.fromJson(_map(json['technology'])),
        governance:
            CorporationGovernanceProfile.fromJson(_map(json['governance'])),
        constitution:
            CorporationConstitutionProfile.fromJson(_map(json['constitution'])),
      );
}

class CorporationIdentity {
  final String id;
  final String name;
  final String admissionPolicy;
  final String charterVersion;
  const CorporationIdentity(
      {required this.id,
      required this.name,
      required this.admissionPolicy,
      required this.charterVersion});
  factory CorporationIdentity.fromJson(Map<String, dynamic> json) =>
      CorporationIdentity(
        id: _text(json['id']),
        name: _text(json['name']),
        admissionPolicy: _text(json['admissionPolicy'], 'UNKNOWN'),
        charterVersion: _text(json['charterVersion'], 'UNKNOWN'),
      );
}

class CorporationMembershipProfile {
  final int memberHouseCount;
  final bool canJoin;
  final String membershipState;
  const CorporationMembershipProfile(
      {required this.memberHouseCount,
      required this.canJoin,
      required this.membershipState});
  factory CorporationMembershipProfile.fromJson(Map<String, dynamic> json) =>
      CorporationMembershipProfile(
        memberHouseCount: _int(json['memberHouseCount']),
        canJoin: json['canJoin'] == true,
        membershipState: _text(json['membershipState'], 'INELIGIBLE'),
      );
}

class CorporationCapacityProfile {
  final String occupiedUnits;
  final String availableUnits;
  final String residentialUnits;
  final String privateProductiveUnits;
  final String publicUnits;
  final String standardBlockUnits;
  final String requiredBlocks;
  final int utilizationBps;
  final String? houseBaseRateUnits;
  final String status;
  const CorporationCapacityProfile(
      {required this.occupiedUnits,
      required this.availableUnits,
      required this.residentialUnits,
      required this.privateProductiveUnits,
      required this.publicUnits,
      required this.standardBlockUnits,
      required this.requiredBlocks,
      required this.utilizationBps,
      required this.houseBaseRateUnits,
      required this.status});
  factory CorporationCapacityProfile.fromJson(Map<String, dynamic> json) =>
      CorporationCapacityProfile(
        occupiedUnits: _text(json['occupiedUnits']),
        availableUnits: _text(json['availableUnits']),
        residentialUnits: _text(json['residentialUnits']),
        privateProductiveUnits: _text(json['privateProductiveUnits']),
        publicUnits: _text(json['publicUnits']),
        standardBlockUnits: _text(json['standardBlockUnits']),
        requiredBlocks: _text(json['requiredBlocks']),
        utilizationBps: _int(json['utilizationBps']),
        houseBaseRateUnits: json['houseBaseRateUnits']?.toString(),
        status: _text(json['status'], 'CURRENT'),
      );
}

class CorporationCapacityFinance {
  final String? houseRevenueUnits;
  final String? earthExpenseUnits;
  final String? marginUnits;
  final String? arrearsUnits;
  const CorporationCapacityFinance(
      {required this.houseRevenueUnits,
      required this.earthExpenseUnits,
      required this.marginUnits,
      required this.arrearsUnits});
  factory CorporationCapacityFinance.fromJson(Map<String, dynamic> json) =>
      CorporationCapacityFinance(
        houseRevenueUnits: _text(json['houseRevenueUnits']),
        earthExpenseUnits: _text(json['earthExpenseUnits']),
        marginUnits: _text(json['marginUnits']),
        arrearsUnits: _text(json['arrearsUnits']),
      );
}

class CorporationAccounts {
  final String treasuryUnits;
  final String operationsUnits;
  final String reserveUnits;
  const CorporationAccounts(
      {required this.treasuryUnits,
      required this.operationsUnits,
      required this.reserveUnits});
  factory CorporationAccounts.fromJson(Map<String, dynamic> json) =>
      CorporationAccounts(
        treasuryUnits: _text(json['treasuryUnits']),
        operationsUnits: _text(json['operationsUnits']),
        reserveUnits: _text(json['reserveUnits']),
      );
}

class CorporationFiscalProfile {
  final String? authorizedUnits;
  final String? committedUnits;
  final String? availableUnits;
  final String? dailyRevenueUnits;
  final String? dailyExpenseUnits;
  const CorporationFiscalProfile(
      {required this.authorizedUnits,
      required this.committedUnits,
      required this.availableUnits,
      required this.dailyRevenueUnits,
      required this.dailyExpenseUnits});
  factory CorporationFiscalProfile.fromJson(Map<String, dynamic> json) =>
      CorporationFiscalProfile(
        authorizedUnits: _text(json['authorizedUnits']),
        committedUnits: _text(json['committedUnits']),
        availableUnits: _text(json['availableUnits']),
        dailyRevenueUnits: _text(json['dailyRevenueUnits']),
        dailyExpenseUnits: _text(json['dailyExpenseUnits']),
      );
}

class CorporationTechnologyProfile {
  final int adoptedCount;
  final int activeResearchCount;
  const CorporationTechnologyProfile(
      {required this.adoptedCount, required this.activeResearchCount});
  factory CorporationTechnologyProfile.fromJson(Map<String, dynamic> json) =>
      CorporationTechnologyProfile(
          adoptedCount: _int(json['adoptedCount']),
          activeResearchCount: _int(json['activeResearchCount']));
}

class CorporationGovernanceProfile {
  final int openProposalCount;
  final List<String> roles;
  final List<String> permissions;
  const CorporationGovernanceProfile(
      {required this.openProposalCount,
      required this.roles,
      required this.permissions});
  factory CorporationGovernanceProfile.fromJson(Map<String, dynamic> json) =>
      CorporationGovernanceProfile(
        openProposalCount: _int(json['openProposalCount']),
        roles: _strings(json['roles']),
        permissions: _strings(json['permissions']),
      );
}

class CorporationConstitutionProfile {
  final Map<String, dynamic> rules;
  final Map<String, dynamic> provenance;
  final List<CorporationPolicy> policies;
  const CorporationConstitutionProfile(
      {required this.rules, required this.provenance, required this.policies});
  factory CorporationConstitutionProfile.fromJson(Map<String, dynamic> json) =>
      CorporationConstitutionProfile(
          rules: _map(json['rules']),
          provenance: _map(json['provenance']),
          policies: (json['policies'] is List
                  ? json['policies'] as List
                  : const [])
              .whereType<Map>()
              .map((row) =>
                  CorporationPolicy.fromJson(Map<String, dynamic>.from(row)))
              .toList());
}

class CorporationPolicy {
  final String ruleCode;
  final dynamic value;
  final String source;
  final String? version;
  final int? effectiveFromGameDay;
  final String? calculationKey;
  final String? policyGroup;
  final String? valueType;
  final String? articleCode;
  const CorporationPolicy({
    required this.ruleCode,
    required this.value,
    required this.source,
    required this.version,
    required this.effectiveFromGameDay,
    required this.calculationKey,
    required this.policyGroup,
    required this.valueType,
    required this.articleCode,
  });
  factory CorporationPolicy.fromJson(Map<String, dynamic> json) =>
      CorporationPolicy(
        ruleCode: _text(json['ruleCode'], 'UNKNOWN_RULE'),
        value: json['value'],
        source: _text(json['source'], 'EARTH'),
        version: json['version']?.toString(),
        effectiveFromGameDay:
            int.tryParse(json['effectiveFromGameDay']?.toString() ?? ''),
        calculationKey: json['calculationKey']?.toString(),
        policyGroup: json['policyGroup']?.toString(),
        valueType: json['valueType']?.toString(),
        articleCode: json['articleCode']?.toString(),
      );
}

Map<String, dynamic> _map(dynamic value) =>
    value is Map ? Map<String, dynamic>.from(value) : <String, dynamic>{};
String _text(dynamic value, [String fallback = '0']) =>
    value?.toString() ?? fallback;
int _int(dynamic value) => int.tryParse(value?.toString() ?? '') ?? 0;
List<String> _strings(dynamic value) =>
    value is List ? value.map((item) => item.toString()).toList() : <String>[];

File: cloudflare/src/corporation-profile.ts
export type CorporationMembershipState = 'MEMBER' | 'PENDING' | 'ELIGIBLE' | 'INELIGIBLE';

export type CorporationPolicy = {
  ruleCode: string;
  value: unknown;
  source: 'EARTH' | 'CORPORATION';
  version: number | string | null;
  effectiveFromGameDay: number | null;
  calculationKey: string | null;
  policyGroup: string | null;
  valueType: string | null;
  articleCode: string | null;
};

export type CorporationProfile = {
  identity: {
    id: string;
    name: string;
    admissionPolicy: string;
    charterVersion: string;
  };
  membership: {
    memberHouseCount: number;
    canJoin: boolean;
    membershipState: CorporationMembershipState;
  };
  capacity: {
    occupiedUnits: string;
    availableUnits: string;
    residentialUnits: string;
    privateProductiveUnits: string;
    publicUnits: string;
    standardBlockUnits: string;
    requiredBlocks: string;
    utilizationBps: number;
    houseBaseRateUnits: string | null;
    status: string;
  };
  capacityFinance: {
    houseRevenueUnits: string | null;
    earthExpenseUnits: string | null;
    marginUnits: string | null;
    arrearsUnits: string | null;
  };
  accounts: {
    treasuryUnits: string;
    operationsUnits: string;
    reserveUnits: string;
  };
  fiscal: {
    authorizedUnits: string | null;
    committedUnits: string | null;
    availableUnits: string | null;
    dailyRevenueUnits: string | null;
    dailyExpenseUnits: string | null;
  };
  technology: {
    adoptedCount: number;
    activeResearchCount: number;
  };
  governance: {
    openProposalCount: number;
    roles: string[];
    permissions: string[];
  };
  constitution: {
    rules: Record<string, unknown>;
    provenance: Record<string, unknown>;
    policies: CorporationPolicy[];
  };
};

export type CorporationDirectoryEntry = {
  id: string;
  name: string;
  admissionPolicy: string;
  memberHouseCount: number;
  incomeTaxBps: number | null;
  salesTaxBps: number | null;
  corporateTaxBps: number | null;
  propertyTaxBps: number | null;
  houseCapacityBaseRateUnits: string | null;
  occupiedCapacityUnits: string;
  standardCapacityUnits: string;
  requiredStandardUnits: string;
  capacityUtilizationBps: number;
  houseCapacityRevenueUnits: string;
  earthCapacityExpenseUnits: string;
  capacityMarginUnits: string;
  capacityStatus: string;
  treasuryUnits: string;
  operationsUnits: string;
  reserveUnits: string;
  technologyCount: number;
  activeResearchCount: number;
  canJoin: boolean;
  membershipState: CorporationMembershipState;
};

type CorporationProfileSource = Record<string, unknown>;

const text = (source: CorporationProfileSource, ...keys: string[]): string => {
  for (const key of keys) {
    if (source[key] != null) return String(source[key]);
  }
  return '0';
};

const optionalText = (source: CorporationProfileSource, ...keys: string[]): string | null => {
  for (const key of keys) {
    if (source[key] != null) return String(source[key]);
  }
  return null;
};

const numberValue = (source: CorporationProfileSource, ...keys: string[]): number => {
  for (const key of keys) {
    if (source[key] != null) return Number(source[key]);
  }
  return 0;
};

const optionalNumber = (source: CorporationProfileSource, ...keys: string[]): number | null => {
  for (const key of keys) {
    if (source[key] != null) return Number(source[key]);
  }
  return null;
};

export function corporationProfileFromSource(source: CorporationProfileSource): CorporationProfile {
  const membershipState = String(source.membershipState ?? source.membership_state ?? 'INELIGIBLE') as CorporationMembershipState;
  const occupiedUnits = text(source, 'occupiedCapacityUnits', 'occupied_capacity_units', 'v5_occupied_capacity');
  const standardBlockUnits = text(source, 'standardCapacityUnits', 'standard_capacity_units', 'v5_standard_territory_capacity');
  const requiredBlocks = text(source, 'requiredStandardUnits', 'required_standard_units', 'v5_required_territory_units');
  const availableUnits = text(source, 'availableCapacityUnits', 'available_capacity_units');
  return {
    identity: {
      id: text(source, 'id'),
      name: text(source, 'name'),
      admissionPolicy: String(source.admissionPolicy ?? source.admission_policy ?? 'UNKNOWN'),
      charterVersion: String(source.charterVersion ?? source.charter_version ?? 'UNKNOWN'),
    },
    membership: {
      memberHouseCount: numberValue(source, 'memberHouseCount', 'member_house_count', 'member_count'),
      canJoin: Boolean(source.canJoin ?? source.can_join),
      membershipState,
    },
    capacity: {
      occupiedUnits,
      availableUnits,
      residentialUnits: text(source, 'residentialUnits', 'residential_units'),
      privateProductiveUnits: text(source, 'privateProductiveUnits', 'private_productive_units'),
      publicUnits: text(source, 'publicUnits', 'public_units'),
      standardBlockUnits,
      requiredBlocks,
      utilizationBps: numberValue(source, 'capacityUtilizationBps', 'capacity_utilization_bps'),
      houseBaseRateUnits: optionalText(source, 'houseCapacityBaseRateUnits', 'house_capacity_base_rate_units'),
      status: String(source.capacityStatus ?? source.capacity_status ?? 'CURRENT'),
    },
    capacityFinance: {
      houseRevenueUnits: optionalText(source, 'houseCapacityRevenueUnits', 'house_capacity_revenue_units'),
      earthExpenseUnits: optionalText(source, 'earthCapacityExpenseUnits', 'earth_capacity_expense_units'),
      marginUnits: optionalText(source, 'capacityMarginUnits', 'capacity_margin_units'),
      arrearsUnits: optionalText(source, 'capacityArrearsUnits', 'capacity_arrears_units'),
    },
    accounts: {
      treasuryUnits: text(source, 'treasuryUnits', 'treasury_units'),
      operationsUnits: text(source, 'operationsUnits', 'operations_units'),
      reserveUnits: text(source, 'reserveUnits', 'reserve_units'),
    },
    fiscal: {
      authorizedUnits: optionalText(source, 'authorizedUnits', 'authorized_units'),
      committedUnits: optionalText(source, 'committedUnits', 'committed_units'),
      availableUnits: optionalText(source, 'availableUnits', 'available_units'),
      dailyRevenueUnits: optionalText(source, 'dailyRevenueUnits', 'daily_revenue_units'),
      dailyExpenseUnits: optionalText(source, 'dailyExpenseUnits', 'daily_expense_units'),
    },
    technology: {
      adoptedCount: numberValue(source, 'technologyCount', 'technology_count'),
      activeResearchCount: numberValue(source, 'activeResearchCount', 'active_research_count'),
    },
    governance: {
      openProposalCount: numberValue(source, 'openProposalCount', 'open_proposal_count'),
      roles: Array.isArray(source.roles) ? source.roles.map(String) : [],
      permissions: Array.isArray(source.permissions) ? source.permissions.map(String) : [],
    },
    constitution: {
      rules: (source.rules ?? source.constitutionRules ?? source.constitution_rules) && typeof (source.rules ?? source.constitutionRules ?? source.constitution_rules) === 'object'
        ? (source.rules ?? source.constitutionRules ?? source.constitution_rules) as Record<string, unknown>
        : {},
      provenance: source.provenance && typeof source.provenance === 'object' ? source.provenance as Record<string, unknown> : {},
      policies: Array.isArray(source.policies) ? source.policies as CorporationPolicy[] : [],
    },
  };
}

export function corporationPoliciesFromConstitution(readModel: Record<string, unknown>): CorporationPolicy[] {
  const rules = readModel.rules && typeof readModel.rules === 'object' ? readModel.rules as Record<string, unknown> : {};
  const provenance = readModel.provenance && typeof readModel.provenance === 'object' ? readModel.provenance as Record<string, unknown> : {};
  const definitions = Array.isArray(readModel.definitions) ? readModel.definitions as Array<Record<string, unknown>> : [];
  const history = Array.isArray(readModel.history) ? readModel.history as Array<Record<string, unknown>> : [];
  return Object.entries(rules).map(([ruleCode, value]) => {
    const source = String(provenance[ruleCode] ?? 'EARTH') as 'EARTH' | 'CORPORATION';
    const definition = definitions.find((row) => String(row.rule_code) === ruleCode);
    const version = history.find((row) => String(row.rule_code) === ruleCode && String(row.authority_type) === source);
    return {
      ruleCode,
      value,
      source,
      version: version?.version == null ? null : version.version as number | string,
      effectiveFromGameDay: version?.effective_from_game_day == null ? null : Number(version.effective_from_game_day),
      calculationKey: definition?.calculation_key == null ? null : String(definition.calculation_key),
      policyGroup: definition?.policy_group == null ? null : String(definition.policy_group),
      valueType: definition?.value_type == null ? null : String(definition.value_type),
      articleCode: definition?.article_code == null ? null : String(definition.article_code),
    };
  });
}

export function corporationDirectoryEntryFromProfile(profile: CorporationProfile, source: CorporationProfileSource = {}): CorporationDirectoryEntry {
  const tax = (ruleCode: string, ...legacyKeys: string[]): number | null => {
    const rule = profile.constitution.rules[ruleCode];
    if (rule != null) return Number(rule);
    return optionalNumber(source, ...legacyKeys);
  };
  return {
    id: profile.identity.id,
    name: profile.identity.name,
    admissionPolicy: profile.identity.admissionPolicy,
    memberHouseCount: profile.membership.memberHouseCount,
    incomeTaxBps: tax('CORPORATION.TAX.INCOME_RATE', 'incomeTaxBps', 'income_tax_bps'),
    salesTaxBps: tax('CORPORATION.TAX.SALES_RATE', 'salesTaxBps', 'sales_tax_bps'),
    corporateTaxBps: tax('CORPORATION.TAX.CORPORATE_RATE', 'corporateTaxBps', 'corporate_tax_bps'),
    propertyTaxBps: tax('CORPORATION.TAX.PROPERTY_RATE', 'propertyTaxBps', 'property_tax_bps'),
    houseCapacityBaseRateUnits: profile.capacity.houseBaseRateUnits,
    occupiedCapacityUnits: profile.capacity.occupiedUnits,
    standardCapacityUnits: profile.capacity.standardBlockUnits,
    requiredStandardUnits: profile.capacity.requiredBlocks,
    capacityUtilizationBps: profile.capacity.utilizationBps,
    houseCapacityRevenueUnits: profile.capacityFinance.houseRevenueUnits ?? '0',
    earthCapacityExpenseUnits: profile.capacityFinance.earthExpenseUnits ?? '0',
    capacityMarginUnits: profile.capacityFinance.marginUnits ?? '0',
    capacityStatus: profile.capacity.status,
    treasuryUnits: profile.accounts.treasuryUnits,
    operationsUnits: profile.accounts.operationsUnits,
    reserveUnits: profile.accounts.reserveUnits,
    technologyCount: profile.technology.adoptedCount,
    activeResearchCount: profile.technology.activeResearchCount,
    canJoin: profile.membership.canJoin,
    membershipState: profile.membership.membershipState,
  };
}
